app/services/archive/unvalidated_user.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+module Archive
+  class UnvalidatedUser < ::BaseService
+    EXCLUDED_ROLES = %w[
+      appropriate_body lead_provider delivery_partner admin finance induction_coordinator npq_participant npq_applicant
+    ].freeze
+
+    def call
+      check_user_can_be_archived!
+
+      data = Archive::UserSerializer.new(user).serializable_hash[:data]
+
+      ActiveRecord::Base.transaction do
+        relic = Archive::Relic.create!(object_type: user.class.name,
+                                       object_id: user.id,
+                                       display_name: user.full_name,
+                                       reason:,
+                                       data:)
+        destroy_user! unless keep_original
+        relic
+      end
+    end
+
+  private
+
+    attr_accessor :user, :reason, :keep_original
+
+    def initialize(user, reason: "unvalidated/undeclared ECTs 2021 or 2022", keep_original: false)
+      @user = user
+      @reason = reason
+      @keep_original = keep_original
+    end
+
+    def check_user_can_be_archived!
+      if users_excluded_roles.any?
+        raise ArchiveError, "User #{user.id} has excluded roles: #{users_excluded_roles.join(',')}"
+      elsif user_has_declarations?
+        raise ArchiveError, "User #{user.id} has non-voided declarations"
+      elsif user_has_eligibility?
+        raise ArchiveError, "User #{user.id} has an eligibility record"
+      elsif user_has_mentees?
+        raise ArchiveError, "User #{user.id} has mentees"
+      end
+    end
+
+    def users_excluded_roles
+      @users_excluded_roles ||= (user.user_roles & EXCLUDED_ROLES)
+    end
+
+    def user_has_declarations?
+      ParticipantDeclaration.not_voided.where(participant_profile_id: user.participant_profiles.select(:id)).any?
+    end
+
+    def user_has_eligibility?
+      ECFParticipantEligibility.where(participant_profile_id: user.participant_profiles.select(:id)).any?
+    end
+
+    def user_has_mentees?
+      return false unless user.user_roles.include? "mentor"
+
+      user.teacher_profile.participant_profiles.mentors.any? do |mentor_profile|
+        InductionRecord.where(mentor_profile:).any?
+      end
+    end
+
+    def destroy_user!
+      user.participant_identities.each do |participant_identity|
+        participant_identity.participant_profiles.each do |participant_profile|
+          destroy_profile_data!(participant_profile)
+        end
+        participant_identity.destroy!
+      end
+      user.teacher_profile.destroy!
+      user.destroy!
```

Comment:
I was wondering if it was possible to avoid deletion of users, and maybe archiving instead as we wanted to avoid deleting those for audit purposes?

Reply:
In this instance these are the ones that Nathan wants deleted, they will be unvalidated, undeclared in the initial batch so shouldn't have much, if any, history.  The policy team have said that these have been chased for months/years via the schools and providers and if needed in the future will be added as new records.